lxd/storage/drivers/driver_zfs_volumes.go

```diff
@@ -797,7 +943,175 @@ func (d *zfs) CreateVolumeFromMigration(vol Volume, conn io.ReadWriteCloser, vol
+		} else {
+			sender = exec.Command("zfs", "send", d.dataset(src, false))
+		}
+
+		var senderErrBuf bytes.Buffer
+		var receiverErrBuf bytes.Buffer
+
+		// Configure the pipes.
+		sender.Stderr = &senderErrBuf
+		receiver.Stdin, _ = sender.StdoutPipe()
+		receiver.Stdout = os.Stdout
+		receiver.Stderr = &receiverErrBuf
+
+		// Run the transfer.
+		err := receiver.Start()
+		if err != nil {
+			return fmt.Errorf("Failed to receive stream: %w", err)
+		}
+
+		err = sender.Run()
+		if err != nil {
+			return fmt.Errorf("Failed to send stream %q: %s: %w", sender.String(), senderErrBuf.String(), err)
+		}
+
+		err = receiver.Wait()
+		if err != nil {
+			return fmt.Errorf("Failed to wait for receiver: %s: %w", receiverErrBuf.String(), err)
+		}
+
+		return nil
+	}
+
+	// This represents the most recent identical snapshot of the source volume and target volume.
+	lastIdenticalSnapshot := targetSnapshots[len(targetSnapshots)-1]
+	_, lastIdenticalSnapshotOnlyName, _ := shared.InstanceGetParentAndSnapshotName(lastIdenticalSnapshot.Name())
+
+	// Rollback target volume to the latest identical snapshot
+	err = d.RestoreVolume(vol, lastIdenticalSnapshotOnlyName, op)
+	if err != nil {
+		return fmt.Errorf("Failed to restore volume: %w", err)
+	}
+
+	// Create all missing snapshots on the target using an incremental stream
+	for i, snap := range srcSnapshots {
+		var originSnap Volume
+
+		if i == 0 {
+			originSnap, err = srcVol.NewSnapshot(lastIdenticalSnapshotOnlyName)
+			if err != nil {
+				return fmt.Errorf("Failed to create new snapshot volume: %w", err)
+			}
+		} else {
+			originSnap = srcSnapshots[i-1]
+		}
+
+		err = transfer(snap, vol, originSnap)
+		if err != nil {
+			// Don't fail here. If it's not possible to perform an optimized refresh, do a generic
+			// refresh instead.
+			d.logger.Debug("Unable to perform an optimized refresh, doing a generic refresh", log.Ctx{"err": err})
+			return genericVFSCopyVolume(d, nil, vol, srcVol, srcSnapshots, true, op)
+		}
+
+		if snap.IsVMBlock() {
+			srcFSVol := snap.NewVMBlockFilesystemVolume()
+			targetFSVol := vol.NewVMBlockFilesystemVolume()
+			originFSVol := originSnap.NewVMBlockFilesystemVolume()
+
+			err = transfer(srcFSVol, targetFSVol, originFSVol)
+			if err != nil {
+				// Don't fail here. If it's not possible to perform an optimized refresh, do a generic
+				// refresh instead.
+				d.logger.Debug("Unable to perform an optimized refresh, doing a generic refresh", log.Ctx{"err": err})
+				return genericVFSCopyVolume(d, nil, vol, srcVol, srcSnapshots, true, op)
+			}
+		}
+	}
+
+	// Create temporary snapshot of the source volume.
+	snapUUID := uuid.New()
+
+	srcSnap, err := srcVol.NewSnapshot(snapUUID)
+	if err != nil {
+		return err
+	}
+
+	err = d.CreateVolumeSnapshot(srcSnap, op)
+	if err != nil {
+		return err
+	}
+
+	latestSnapVol := srcSnapshotsAll[len(srcSnapshotsAll)-1]
+
+	err = transfer(srcSnap, vol, latestSnapVol)
+	if err != nil {
```

Comment:
This seems a bit, umm, "hand wavy", basically if anything goes wrong, then fallback to generic copy method.
I would rather see that `transfer` return a specific error that indicates that optimized transfer is not available which triggers the fallback, and otherwise fail with an error so that genuine errors (like network issues) are not covered up.

Reply:
Done